game/scripts/vscripts/components/abilities/level.lua

```diff
@@ -16,43 +16,33 @@ function AbilityLevels:FilterAbilityUpgradeOrder (keys)
   end
 
   -- Ability hero level requirements
-  local basicLevel5Req = 28
-  local basicLevel6Req = 40
-  local ultimateLevel4Req = 37
-  local ultimateLevel5Req = 49
+  local basic_reqs = {0, 0, 0, 0, 28, 40}
+  local ultimate_reqs = {0, 0, 0, 0, 37, 49}
 
   local ability = EntIndexToHScript(keys.entindex_ability)
   local player = PlayerResource:GetPlayer(keys.issuer_player_id_const)
   local hero = EntIndexToHScript(keys.units["0"])
   local heroLevel = hero:GetLevel()
   local abilityLevel = ability:GetLevel()
   local abilityType = ability:GetAbilityType()
-  local abilityCanUpgrade = true
-  local requriedHeroLevel = -1
 
   if abilityType == 0 then -- Ability is DOTA_ABILITY_TYPE_BASIC
-    if abilityLevel >= 4 and heroLevel < basicLevel5Req then
-      abilityCanUpgrade = false
-      requiredHeroLevel = basicLevel5Req
-    elseif abilityLevel >= 5 and heroLevel < basicLevel6Req then
-      abilityCanUpgrade = false
-      requiredHeroLevel = basicLevel6Req
-    end
+    req_table = basic_reqs
   elseif abilityType == 1 then -- Ability is DOTA_ABILITY_TYPE_ULTIMATE
-    if abilityLevel >= 3 and heroLevel < ultimateLevel4Req then
-      abilityCanUpgrade = false
-      requiredHeroLevel = ultimateLevel4Req
-    elseif abilityLevel >= 4 and heroLevel < ultimateLevel5Req then
-      abilityCanUpgrade = false
-      requiredHeroLevel = ultimateLevel5Req
-    end
+    req_table = ultimate_reqs
   end
-
-  if abilityCanUpgrade then
+  
+  if abilityLevel > #req_table then
+    requirement = req_table[#req_table]
+  else
+    requirement = req_table[abilityLevel]
```

Comment:
Should be `abilityLevel + 1` because the function acts before the ability upgrade, meaning that when trying to go from level 4 to 5 of a basic ability, `abilityLevel` would be 4